pydicom/charset.py

```diff
@@ -91,24 +107,33 @@ def decode_string(value, encodings):
                                             escape_codes['iso_ir_58']))
 
     if use_python_handling:
-        parts = [value]
+        values = [value]
     else:
-        # Each part of the value that starts with an escape sequence is
-        # decoded separately using the corresponding encoding.
+        # Each part of the value that starts with an escape sequence
+        # or a delimiter as defined in PS3.5, section 6.1.2.5.3 is decoded
+        # separately. If it starts with an escape sequence, the
+        # corresponding encoding is used, otherwise the first encoding.
         # See PS3.5, 6.1.2.4 and 6.1.2.5 for the use of code extensions.
-        parts = [ESC + part for part in value.split(ESC) if part]
-        # the first part may not start with an escape sequence
-        if not value.startswith(ESC):
-            parts[0] = parts[0][1:]
+        #
+        # The following regex splits the value into these parts, by matching
+        # substrings at line start not containing any of the delimiters,
+        # and substrings starting with a delimiter and not containing other
+        # delimiters.
+        regex = b'(^[^' + delims + b']+|[' + delims + b'][^' + delims + b']*)'
+        values = re.findall(regex, value)
+
     result = u''
 
-    for part in parts:
+    for part in values:
         if part.startswith(ESC):
             for enc in list(encodings) + ['iso8859']:
                 if enc in escape_codes and part.startswith(escape_codes[enc]):
```

Comment:
What happens if enc is not in escape_codes?  Maybe I missed some discussion about silently ignoring this?'
Oh wait ... looks like the 'else:' is in line with the 'for' - I think it is meant to be the else clause of the if?

Reply:
Also, I wonder ... how about a function `find_encoding(part)` to remove some of this logic from the loop.  Since the escape codes are deterministic, the function could use a reverse version of the escape_codes dict.  i.e. escape codes are either ESC + 2 bytes, or ESC + "$" + 2 bytes, look up the first 2 or three bytes and get the encoding directly without a loop.   Still have to deal with unknown escape codes, or ones not matching the `encodings` list.

Reply:
Yes, I'm not happy with this part either - I will change this.

Reply:
I did a bit of a simplification, though not as much as I would have liked.